### ISSUE TITLE
Question, Entity: How to express attribute which is another Hanami::Entity type

### ISSUE THREAD

**Issue by thhermansen** [hanami/model#356]
In [the docs for Entity](http://hanamirb.org/guides/models/entities/) you have the following example

```ruby
# lib/bookshelf/entities/user.rb
class User < Hanami::Entity
  EMAIL_FORMAT = /\@/

  attributes do
    attribute :id,         Types::Int
    attribute :name,       Types::String
    attribute :email,      Types::String.constrained(format: EMAIL_FORMAT)
    attribute :age,        Types::Int.constrained(gt: 18)
    attribute :codes,      Types::Collection(Types::Coercible::Int)
    attribute :comments,   Types::Collection(Comment)
    attribute :created_at, Types::Time
    attribute :updated_at, Types::Time
  end
end
```

The example shows that you can have an array of Comments, where Comment is a `Hanami::Entity`.

### The question

How would you do it if you where to express that a User had _one_ comment?

In other words, something like `attribute :comment, Comment`. If you do this you get an error (`undefined method `default?' for <class name here>`) since Comment isn't something which plays along with dry-types.

### What I have done to try figure it out

I have taken a look at how the `Types::Collection` is implemented and tried to do it like: `attribute :comment, Types::Schema::CoercibleType.new(Comment)`. I know this is an internal API, but I wanted to see if it worked as I though and it kinda did.

Now you can do something like:

```ruby
comment = Comment.new text: 'My comment'
user = User.new comment: comment
```

but when you ask `user.comment` you get back a hash `{text: 'My comment'}` and an instance of Comment.

I have tracked this down to [this code](https://github.com/hanami/model/blob/master/lib/hanami/entity/schema.rb#L207-L211) where the attributes it receives is deeply symbolized. Since Hanami::Entity [responds to `to_hash`](https://github.com/hanami/model/blob/8775b815edfaef38a1d98cdd83704bbcf2c553f9/lib/hanami/entity.rb#L190) it is made in to a hash during User's initialization.

------------------------

Is there any way of doing this which I have missed?

**Comment by solnic** [comment]
I'd say this is a bug because `attribute :comment, Comment` should Just Work™; or maybe it's a missing feature, something like `attribute :comment, Types::Entity(Comment)` that would create a coercible type. Not sure.

**Comment by thhermansen** [comment]
Thanks for your reply! :-)

Yes, I though of maybe doing a pull request with `Types::Entity`, so it mirrors `Types::Collection`. They schema is doing the correct thing thing as far as I see when I decorated the entity class with `Types::Schema::CoercibleType`. The showstopper became [symbolize!](https://github.com/hanami/model/blob/master/lib/hanami/entity/schema.rb#L207-L211) which makes a hash out of every value in attributes responding to `to_hash`. I'm not sure if we really must deeply symbolize keys here so I didn't go forward with the pull request.

----------------

As an experiment I tried to `undef_method :to_hash` on my entity and it then worked as expected. 

However I I ran in to trouble again when I passed that entity with no `to_hash` to `Repository#create` as data. I think I read it is preferable to pass hash and not entity to #create/update, but we do have a bit of legacy code, pluss I like the idea to pass the frozen-hash-like-hanami-entity to the repository as data as well as a hash.

-------------

 I guess it could be a missing feature as one-to-one association does not exist yet, [ref. docs](http://hanamirb.org/guides/models/associations/).
